### PR TITLE
[pushbullet] Add link and file push type support

### PR DESCRIPTION
**Pull request by jsetton**
This change adds link and file push type support as defined in the [Pushbullet API documentation](https://docs.pushbullet.com/#create-push).

This includes the ability to send messages attaching the content of an image URL, a local file or an Image item state.

### CODE REVIEW

File: bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java [hunk 1]
```diff
@@ -87,9 +87,35 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
         logger.debug("Start initializing!");
-        config = getConfigAs(PushbulletConfiguration.class);
+        PushbulletConfiguration config = getConfigAs(PushbulletConfiguration.class);
+
+        httpClient.setConfiguration(config);
+
+        if (config.getToken().isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Undefined token.");
+            return;
+        }
```
Comment (lsiepel) [inline bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:97]: Minor, better to first check and then set
```suggestion
        if (config.getToken().isEmpty()) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Undefined token.");
            return;
        }
        
        httpClient.setConfiguration(config);
```

File: bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java [hunk 2]
```diff
@@ -87,9 +87,35 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+
```
Comment (lsiepel) [inline bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:98]:
```suggestion

    updateStatus(ThingStatus.UNKNOWN);

```

File: bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java [hunk 3]
```diff
@@ -87,9 +87,35 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        User user = httpClient.executeRequest(API_ENDPOINT_USERS_ME, User.class);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:99]: `initialize()` should return fast. This call is blocking. It should be run in async. 

I think that from this line to the end of the block, could be extracted to a seperate private method that can be called from a `scheduler.execute` task

**Review by lsiepel — CHANGES_REQUESTED** [review]
In general it looks very good, nice additions! Left some comments to look at.

File: bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/PushbulletHttpClient.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+     *
+     * @param url the upload url
+     * @param data the file data
+     * @return true if response status code 204
+     */
+    public boolean uploadFile(String url, RawType data) {
+        MultiPartContentProvider content = new MultiPartContentProvider();
+        content.addFieldPart("file", new BytesContentProvider(data.getMimeType(), data.getBytes()), null);
+
+        Request request = newRequest(url).method(HttpMethod.POST).content(content);
```
[the remark this answers is not in the document]
Comment (jsetton) [reply bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/PushbulletHttpClient.java:130]: The timeout parameter is set in the `newRequest` wrapper method.

File: bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java [hunk 5]
```diff
@@ -86,139 +88,270 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        }
 
-        // Name and Token are both "required", so set the Thing immediately ONLINE.
-        updateStatus(ThingStatus.ONLINE);
+        httpClient.setConfiguration(config);
 
-        logger.debug("Finished initializing!");
+        scheduler.execute(() -> retrieveAccountInfo());
+
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:104]: For reference: https://www.openhab.org/docs/concepts/things.html#status-details
```suggestion
        updateStatus(ThingStatus.UNKNOWN);
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Last comment, otherwise LGTM. Thanks for addressing it quickly!

**Comment by jsetton** [comment]
I was wondering if the channel configuration should be removed. There is no implementation in the binding supporting them and is confusing.

Same goes with the items and sitemap examples at the end of the documentation.

File: bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -22,8 +22,8 @@
 			</parameter>
 
 			<parameter name="token" type="text" required="true">
-				<label>Token</label>
-				<description>Token as obtained from the server</description>
+				<label>Access Token</label>
+				<description>Access token obtained from the account settings page</description>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml:26]: This would aslo require a i18n update

**Comment by lsiepel** [comment]
> I was wondering if the channel configuration should be removed. There is no implementation in the binding supporting them and is confusing. Same goes with the items and sitemap examples at the end of the documentation.
> 
> As far as the thing configuration, I think that the `apiUrlBase` parameter should be hardcoded. There is no way to run your own Pushbullet server as far as I know. And the `name` parameter isn't used as well.

I prefer to seperate PR's for these kind of clean-up.

File: bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -16,19 +16,19 @@
 
 		<config-description>
 			<parameter name="name" type="text" required="false">
-				<label>Name</label>
-				<description>Explicit Name of Bot, if wanted</description>
+				<label>@text/thing-type.config.pushbullet.bot.name.label</label>
+				<description>@text/thing-type.config.pushbullet.bot.name.description</description>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml:20]: Help me out here, what is the purpose of replacing the default labels and descriptions with these i18n placeholders?

Comment (jsetton) [reply bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml:20]: I am mirrored what was done previously with the other i18n placeholders for this binding such as the `addon.xml` or addon resources.

I don't usually do this but I wanted to keep the consistency. I can revert that change if necessary.

Comment (lsiepel) [reply bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml:20]: Yes please revert.

Comment (jsetton) [reply bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml:20]: Should I update `addon.xml` as well?

Comment (jsetton) [reply bundles/org.openhab.binding.pushbullet/src/main/resources/OH-INF/thing/thing-types.xml]: I ended updating that file as well for consistency.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM.

Do you need some time to finish tests and validate no regressions happened due to all the recent commits? I'll merge otherwise.

**Comment by jsetton** [comment]
> Do you need some time to finish tests and validate no regressions happened due to all the recent commits?

I already tested that I could push a note, link and file successfully with the last changes I made. So the changes can be merged. Thanks.

**Comment by jsetton** [comment]
> I prefer to seperate PR's for these kind of clean-up.

#17482

**Pull request merged**